### ISSUE TITLE
Valid language codes do not match settings example documentation

### ISSUE THREAD

**Issue by dbyte** [dmgbuild/dmgbuild#57]
Docs in example of settings.py say
```
# Recognized language names are:
#
#  af_ZA, ar, be_BY, bg_BG, bn, bo, br, ca_ES, cs_CZ, cy, da_DK, de_AT, de_CH,
#  de_DE, dz_BT, el_CY, el_GR, en_AU, en_CA, en_GB, en_IE, en_SG, en_US, eo,
#  es_419, es_ES, et_EE, fa_IR, fi_FI, fo_FO, fr_001, fr_BE, fr_CA, fr_CH,
#  fr_FR, ga-Latg_IE, ga_IE, gd, grc, gu_IN, gv, he_IL, hi_IN, hr_HR, hu_HU,
#  hy_AM, is_IS, it_CH, it_IT, iu_CA, ja_JP, ka_GE, kl, ko_KR, lt_LT, lv_LV,
#  mk_MK, mr_IN, mt_MT, nb_NO, ne_NP, nl_BE, nl_NL, nn_NO, pa, pl_PL, pt_BR,
#  pt_PT, ro_RO, ru_RU, se, sk_SK, sl_SI, sr_RS, sv_SE, th_TH, to_TO, tr_TR,
#  uk_UA, ur_IN, ur_PK, uz_UZ, vi_VN, zh_CN, zh_TW
```
But running this with with a config like

```
pathToGermanLicense = 'xxxxxx/EULA.rtf'

license = {
    "default-language": "de_DE",
    "licenses": {
        # For each language, the text of the license.  This can be plain text,
        # RTF (in which case it must start "{\rtf1"), or a path to a file
        # containing the license text.  If you're using RTF,
        # watch out for Python escaping (or read it from a file).
        "de_DE": pathToGermanLicense,
        "de_CH": pathToGermanLicense,
    },
}
```
results in

`File "xxxx/dmgbuild/licensing.py", line 209, in build_license
    raise Exception(
Exception: Unknown language 'de_CH'. Valid languages are: da_DK, de_DE, en_AU, en_GB, en_NZ, en_US, es_ES, fr_CA, fr_FR, it_IT, ja_JP, nb_NO, nl_BE, nl_NL, pt_BR, pt_PT, sv_SE, zh_CN, zh_TW`

Tested on
**macOS 13.2.1
Python 3.11.1**

**Comment by freakboy3742** [comment]
Thanks for the report; looks like a case where the recent change to using udifrez to build translations sets has introduced a discrepancy.

This should be a relatively straightforward fix - either the `language_names_map` in licensing.py needs to be extended to include other common language analogs, or the language map lookup on L209 (where the exception is being raised) needs to be extended to allow for "fallback" languages - in your case, `de_CH` can fall back to `de_DE`. 

I'd suggest we modify the keys in `language_name_map` to use a wildcard when the region doesn't affect the translation (e.g., `en_*`, `de_*`); the have the language lookup fall back to a lookup on `f'{language.split("_")[0]}_*'`. We need to keep the full language lookup for languages like `zh_CN`/`zh_TW`, where the region code is significant; using the wildcard in the key lookup means the error message that is generated can stay the same.

We can then update the docs to match the new language list.

**Comment by dbyte** [comment]
Thanks for your reply.

Note that there also is one more Bug in license.py which seems to be the reason for #56 
Each separator between the button titles must carry the **length** of the localized title. You hard-coded them, which is wrong.

`xml["STR#"].append(
            {
                "Attributes": "0x0000",
                "Data": (
                    b"\x00\x06\x07"
                    + buttons[0]
                    + b"\x05"
                    + buttons[1]
                    + b"\x08"
                    + buttons[2]
                    + b"\x05"
                    + buttons[3]
                    + b"\x07"
                    + buttons[4]
                    + b"{"
                    + buttons[5]
                ),
                "ID": "5002",
                "Name": language_name,
            }
        )`

**Comment by freakboy3742** [comment]
Thanks for that debugging info; care to try your hand at a pull request?

**Comment by dbyte** [comment]
> care to try your hand at a pull request?

Yes, I am already in the process of trying to find a solution that is as straight-forward as possible.
My favorite approach would be to have an additional utility class or button class or both. I would then also add some inline-typing and some unit tests for their functionality - the project seems a bit bare-bones to me in that regard.

My linter shows some heavy noise for the project 🙂... Do you use a linter at all - are there any specifications on your part?

If you have some patience, about 2 weeks, I'll be very happy to tackle this, thank you for the suggestion!

**Comment by freakboy3742** [comment]
> > care to try your hand at a pull request?
> 
> Yes, I am already in the process of trying to find a solution that is as straight-forward as possible. My favorite approach would be to have an additional utility class or button class or both. I would then also add some inline-typing and some unit tests for their functionality - the project seems a bit bare-bones to me in that regard.

I won't disagree there. I inherited the project from someone else who was no longer able to maintain it; the state of testing reflects what was in place when I got here. Ordinarily, I prefer much more rigorous testing, but I don't have the spare time to build a test suite from scratch.
 
> My linter shows some heavy noise for the project 🙂... Do you use a linter at all - are there any specifications on your part?

Yes - we use both flake8 and black. These are configured as pre-commit hooks, and are checked in CI.

**Comment by ABBAPOH** [comment]
Hello, any luck on fixing the issue? I've encountered the problem when trying to use ko_KR and ru_RU licenses.
Also, it seems that hdiutil does not like non-ascii characters when I add localised language name to the language_names_map - any idea how to fix this?

**Comment by ABBAPOH** [comment]
Also, it seems that resource Id might depend on the language https://github.com/argv-minus-one/dmg-license/blob/master/language-info-generator/Languages.tsv.
Also, this JS tool might be an inspiration - it inserts license and buttons correctly (at least, for Japaneese)

**Comment by ABBAPOH** [comment]
At least, fix for buttons length https://github.com/dmgbuild/dmgbuild/pull/61
